jinahub/indexers/indexer/LMDBIndexer/__init__.py

```diff
@@ -0,0 +1,163 @@
+import os
+import pickle
+from typing import Dict
+
+import lmdb
+from jina import Executor, Document, DocumentArray, requests
+
+from jina_commons import get_logger
+from jina_commons.indexers.dump import export_dump_streaming
+
+
+class LMDBIndexer(Executor):
+    """An lmbd-based DBMS Indexer for Jina
+
+    For more information on lmdb check their documentation: https://lmdb.readthedocs.io/en/release/
+    """
+
+    def __init__(
+        self,
+        map_size: int = 10485760,  # in bytes, 10 MB
+        default_traversal_path: str = 'r',
+        *args,
+        **kwargs,
+    ):
+        super().__init__(*args, **kwargs)
+        self.default_traversal_path = default_traversal_path
+        self.file = os.path.join(self.workspace, 'db.lmdb')
+        if not os.path.exists(self.workspace):
+            os.makedirs(self.workspace)
+        # db for internal usage
+        self.db = 'db'
+        self.logger = get_logger(self)
+        # see https://lmdb.readthedocs.io/en/release/#environment-class for usage
+        self.env = lmdb.Environment(
+            self.file,
+            map_size=map_size,
+            subdir=False,
+            readonly=False,
+            metasync=True,
+            sync=True,
+            map_async=False,
+            mode=493,
+            create=True,
+            readahead=True,
+            writemap=False,
+            meminit=True,
+            max_readers=126,
+            max_dbs=0,  # means only one db
+            max_spare_txns=1,
+            lock=True,
+        )
+
+    @requests(on='/index')
+    def index(self, docs: DocumentArray, parameters: Dict = None, **kwargs):
+        """Add entries to the index
+
+        :param docs: the documents to add
+        :param parameters: parameters to the request
+        """
+        if parameters is None:
+            parameters = {}
+
+        trav_path = parameters.get('traversal_path', self.default_traversal_path)
+        with self.env.begin(write=True) as t:
+            for d in docs.traverse_flat(trav_path):
+                t.put(d.id.encode(), d.SerializeToString())
+
+    @requests(on='/update')
+    def update(self, docs: DocumentArray, parameters: Dict = None, **kwargs):
+        """Update entries from the index by id
+
+        :param docs: the documents to update
+        :param parameters: parameters to the request
+        """
+        if parameters is None:
+            parameters = {}
+        trav_path = parameters.get('traversal_path', self.default_traversal_path)
+        with self.env.begin(write=True) as t:
+            for d in docs.traverse_flat(trav_path):
+                t.replace(d.id.encode(), d.SerializeToString())
+
+    @requests(on='/delete')
+    def delete(self, docs: DocumentArray, parameters: Dict = None, **kwargs):
```

Comment:
typehint is not accurate, but actually it will never be `None`

Reply:
Updated